src/service/url-replacements-impl.js

```diff
@@ -74,6 +74,22 @@ export class GlobalVariableSource extends VariableSource {
     /** @const {!./ampdoc-impl.AmpDoc} */
     this.ampdoc = ampdoc;
 
+    // A meta[name="amp-action-whitelist"] tag, if present, contains,
+    // in its content attribute, a whitelist of actions on the special AMP target.
+    if (this.ampVariableSubstitutionWhitelist_ === undefined
+      && this.ampdoc.getRootNode() && this.ampdoc.getRootNode().head) {
+      const meta =
+        this.ampdoc.getRootNode().head
+            .querySelector('meta[name="amp-variable-substitution-whitelist"]');
+
+      // Cache the whitelist of allowed AMP actions (if provided).
+      if (meta) {
+        this.ampVariableSubstitutionWhitelist_ =
```

Comment:
This field needs JSDoc.

It might be better to move the initialization logic to a helper method and use `null` as the default value for this field. `?string` is usually more idiomatic than `string|undefined`.